daemon/api_themes.go

```diff
@@ -0,0 +1,247 @@
+	status = make(map[string]themeStatus, len(themes))
+
+	for _, theme := range themes {
+		// Skip duplicates
+		if _, ok := status[theme]; ok {
+			continue
+		}
+		if strutil.SortedListContains(installed, theme) {
+			status[theme] = themeInstalled
+			continue
+		}
+		status[theme] = themeUnavailable
+		for _, name := range themePackageCandidates(prefix, theme) {
+			var info *snap.Info
+			if info, err = theStore.SnapInfo(ctx, store.SnapSpec{Name: name}, user); err == store.ErrSnapNotFound {
+				continue
+			} else if err != nil {
+				return nil, err
+			}
+			// Only mark the theme as available if it has
+			// been published to the stable channel.
+			if info.Channel == "stable" {
+				status[theme] = themeAvailable
+				toInstall[name] = true
+			}
+			break
+		}
+	}
+	return status, nil
+}
+
+func getThemeStatus(ctx context.Context, c *Command, user *auth.UserState, gtkThemes, iconThemes, soundThemes []string) (status themeStatusResponse, toInstall []string, err error) {
+	installedGtk, installedIcon, installedSound, err := getInstalledThemes(c.d)
+	if err != nil {
+		return themeStatusResponse{}, nil, err
+	}
+
+	theStore := getStore(c)
+	candidates := make(map[string]bool)
+	if status.GtkThemes, err = getThemeStatusForPrefix(ctx, theStore, user, "gtk-theme-", gtkThemes, installedGtk, candidates); err != nil {
+		return themeStatusResponse{}, nil, err
+	}
+	if status.IconThemes, err = getThemeStatusForPrefix(ctx, theStore, user, "icon-theme-", iconThemes, installedIcon, candidates); err != nil {
+		return themeStatusResponse{}, nil, err
+	}
+	if status.SoundThemes, err = getThemeStatusForPrefix(ctx, theStore, user, "sound-theme-", soundThemes, installedSound, candidates); err != nil {
+		return themeStatusResponse{}, nil, err
+	}
+	toInstall = make([]string, 0, len(candidates))
+	for pkg := range candidates {
+		toInstall = append(toInstall, pkg)
+	}
+	sort.Strings(toInstall)
+
+	return status, toInstall, nil
+}
+
+func checkThemes(c *Command, r *http.Request, user *auth.UserState) Response {
+	ctx := store.WithClientUserAgent(r.Context(), r)
+	q := r.URL.Query()
+	status, _, err := getThemeStatus(ctx, c, user, q["gtk-theme"], q["icon-theme"], q["sound-theme"])
+	if err != nil {
+		return InternalError("cannot get theme status: %s", err)
+	}
+
+	return SyncResponse(status, nil)
+}
+
+type themeInstallReq struct {
+	GtkThemes   []string `json:"gtk-themes"`
+	IconThemes  []string `json:"icon-themes"`
+	SoundThemes []string `json:"sound-themes"`
+}
+
+func installThemes(c *Command, r *http.Request, user *auth.UserState) Response {
+	decoder := json.NewDecoder(r.Body)
+	var req themeInstallReq
+	if err := decoder.Decode(&req); err != nil {
+		return BadRequest("cannot decode request body: %v", err)
+	}
+
+	ctx := store.WithClientUserAgent(r.Context(), r)
+	_, toInstall, err := getThemeStatus(ctx, c, user, req.GtkThemes, req.IconThemes, req.SoundThemes)
```

Comment:
this would then use the available-only flag to themeStatus

Reply:
I renamed it to `themeStatusAndMissingSnaps`.  A bit more verbose, but describes what the function does.